# PCA Excercise (Core)

### Imports

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, ConfusionMatrixDisplay
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.neighbors import KNeighborsClassifier

pd.set_option('display.max_columns',200)
pd.set_option("display.max_info_rows", 800)
pd.set_option('display.max_info_columns',800)

from sklearn import set_config
set_config(transform_output='pandas')

import warnings
warnings.filterwarnings("ignore")

### Functions

In [2]:
def classification_metrics(y_true, y_pred, label='',
                           output_dict=False, figsize=(8,4),
                           normalize='true', cmap='Blues',
                           colorbar=False):
    # Get the classification report
    report = classification_report(y_true, y_pred)
    ## Print header and report
    header = "-"*70
    print(header, f" Classification Metrics: {label}", header, sep='\n')
    print(report)
    ## CONFUSION MATRICES SUBPLOTS
    fig, axes = plt.subplots(ncols=2, figsize=figsize)
    # create a confusion matrix  of raw counts
    ConfusionMatrixDisplay.from_predictions(y_true, y_pred,
                normalize=None, cmap='gist_gray', colorbar=colorbar,
                ax = axes[0],);
    axes[0].set_title("Raw Counts")
    # create a confusion matrix with the test data
    ConfusionMatrixDisplay.from_predictions(y_true, y_pred,
                normalize=normalize, cmap=cmap, colorbar=colorbar,
                ax = axes[1]);
    axes[1].set_title("Normalized Confusion Matrix")
    # Adjust layout and show figure
    fig.tight_layout()
    plt.show()
    # Return dictionary of classification_report
    if output_dict==True:
        report_dict = classification_report(y_true, y_pred, output_dict=True)
        return report_dict

In [3]:
def evaluate_classification(model, X_train, y_train, X_test, y_test,
                         figsize=(6,4), normalize='true', output_dict = False,
                            cmap_train='Blues', cmap_test="Reds",colorbar=False):
    # Get predictions for training data
    y_train_pred = model.predict(X_train)
    # Call the helper function to obtain regression metrics for training data
    results_train = classification_metrics(y_train, y_train_pred, #verbose = verbose,
                                     output_dict=True, figsize=figsize,
                                         colorbar=colorbar, cmap=cmap_train,
                                     label='Training Data')
    print()
    # Get predictions for test data
    y_test_pred = model.predict(X_test)
    # Call the helper function to obtain regression metrics for test data
    results_test = classification_metrics(y_test, y_test_pred, #verbose = verbose,
                                  output_dict=True,figsize=figsize,
                                         colorbar=colorbar, cmap=cmap_test,
                                    label='Test Data' )
    if output_dict == True:
        # Store results in a dataframe if ouput_frame is True
        results_dict = {'train':results_train,
                    'test': results_test}
        return results_dict

### Data Prep

In [4]:
df = pd.read_csv('data/NHANES_Diet_Diabetes_Risk.xls.csv')
df = df.set_index("SEQN")
df.head()

WTDRD1       WTDR2D  DR1DRSTZ  DR1EXMER  DRABF  DRDINT  DR1DBIH  \
SEQN                                                                          
73557  16888.32786  12930.89065         1      49.0    2.0     2.0      6.0   
73558  17932.14387  12684.14887         1      59.0    2.0     2.0      4.0   
73559  59641.81293  39394.23671         1      49.0    2.0     2.0     18.0   
73561  59052.35703  39004.89299         1      63.0    2.0     2.0     18.0   
73562  49890.82866      0.00000         1      49.0    2.0     1.0     11.0   

       DR1DAY  DR1LANG  DR1MNRSP  DR1HELPD  DBQ095Z  DBD100  DRQSPREP  DR1STY  \
SEQN                                                                            
73557     2.0      1.0       1.0      13.0      3.0     2.0       4.0     2.0   
73558     1.0      1.0       1.0      13.0      1.0     2.0       3.0     1.0   
73559     6.0      1.0       1.0      13.0      1.0     1.0       2.0     2.0   
73561     1.0      1.0       1.0      13.0      4.0     NaN       4.0     2.0   
73562     3.0      1.0       1.0      13.0      1.0     3.0       3.0     1.0   

       DR1SKY  DRQSDIET  DRQSDT1  DRQSDT2  DRQSDT3  DRQSDT4  DRQSDT5  DRQSDT6  \
SEQN                                                                            
73557     NaN       2.0      NaN      NaN      NaN      NaN      NaN      NaN   
73558     1.0       2.0      NaN      NaN      NaN      NaN      NaN      NaN   
73559     NaN       1.0      NaN      NaN      NaN      NaN      NaN      NaN   
73561     NaN       1.0      NaN      NaN      NaN      NaN      NaN      NaN   
73562     1.0       2.0      NaN      NaN      NaN      NaN      NaN      NaN   

       DRQSDT7  DRQSDT8  DRQSDT9  DRQSDT10  DRQSDT11  DRQSDT12  DRQSDT91  \
SEQN                                                                       
73557      NaN      NaN      NaN       NaN       NaN       NaN       NaN   
73558      NaN      NaN      NaN       NaN       NaN       NaN       NaN   
73559      7.0      NaN      NaN       NaN       NaN       NaN       NaN   
73561      NaN      NaN      NaN       NaN       NaN       NaN      91.0   
73562      NaN      NaN      NaN       NaN       NaN       NaN       NaN   

       DR1TNUMF  DR1TKCAL  DR1TPROT  DR1TCARB  DR1TSUGR  DR1TFIBE  DR1TTFAT  \
SEQN                                                                          
73557      11.0    1574.0     43.63    239.59    176.47      10.8     52.81   
73558       8.0    5062.0    338.13    423.78     44.99      16.7    124.29   
73559      27.0    1743.0     64.61    224.39    102.90       9.9     65.97   
73561       7.0    1421.0     55.24    178.20     87.78      12.3     55.36   
73562      17.0    1785.0     55.11    189.59     81.75      22.6     93.92   

       DR1TSFAT  DR1TMFAT  DR1TPFAT  DR1TCHOL  DR1TATOC  DR1TATOA  DR1TRET  \
SEQN                                                                         
73557    17.819    18.493     8.829     209.0      3.92       0.0     70.0   
73558    53.408    35.481    20.505    2584.0     20.94       0.0   1477.0   
73559    25.263    20.902    12.953      88.0      8.87       0.0    616.0   
73561     4.479    26.216     1.263      41.0     53.17       0.0    738.0   
73562    22.155    40.013    23.550     534.0     12.30       0.0    272.0   

       DR1TVARA  DR1TACAR  DR1TBCAR  DR1TCRYP  DR1TLYCO  DR1TLZ  DR1TVB1  \
SEQN                                                                       
73557     110.0      60.0     304.0     278.0    1398.0   430.0    1.057   
73558    1488.0       0.0     145.0      12.0       0.0   899.0    4.178   
73559     655.0      21.0     449.0      35.0     515.0   300.0    1.648   
73561     863.0       0.0    1476.0       0.0       0.0     0.0    1.484   
73562     355.0     172.0     776.0     300.0    2339.0  2101.0    1.227   

       DR1TVB2  DR1TNIAC  DR1TVB6  DR1TFOLA  DR1TFA  DR1TFF  DR1TFDFE  \
SEQN                                                                    
73557    

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6643 entries, 73557 to 83729
Data columns (total 168 columns):
 #    Column         Dtype  
---   ------         -----  
 0    WTDRD1         float64
 1    WTDR2D         float64
 2    DR1DRSTZ       int64  
 3    DR1EXMER       float64
 4    DRABF          float64
 5    DRDINT         float64
 6    DR1DBIH        float64
 7    DR1DAY         float64
 8    DR1LANG        float64
 9    DR1MNRSP       float64
 10   DR1HELPD       float64
 11   DBQ095Z        float64
 12   DBD100         float64
 13   DRQSPREP       float64
 14   DR1STY         float64
 15   DR1SKY         float64
 16   DRQSDIET       float64
 17   DRQSDT1        float64
 18   DRQSDT2        float64
 19   DRQSDT3        float64
 20   DRQSDT4        float64
 21   DRQSDT5        float64
 22   DRQSDT6        float64
 23   DRQSDT7        float64
 24   DRQSDT8        float64
 25   DRQSDT9        float64
 26   DRQSDT10       float64
 27   DRQSDT11       float64
 28   DRQSDT12   

In [6]:
df.dtypes.value_counts()

float64    166
int64        2
dtype: int64

In [7]:
df.isna().sum().sum()

432597

In [8]:
# Define Target, X, Y
target = 'Diabetes_Risk'
y = df[target].copy()
X = df.drop(columns=target).copy()

In [9]:
# Train/Test Split
X_train,X_test,y_train,y_test = train_test_split(X,y, random_state=321)
X_train.head()

WTDRD1        WTDR2D  DR1DRSTZ  DR1EXMER  DRABF  DRDINT  DR1DBIH  \
SEQN                                                                            
74781  34584.687590      0.000000         1      49.0    2.0     1.0     20.0   
76977  15205.564340  24452.373790         1       2.0    2.0     2.0      9.0   
79530   4940.540826  13796.381440         1      49.0    2.0     2.0     23.0   
74171   5061.660951   3885.556268         1      49.0    2.0     2.0     11.0   
83046  16431.835630  12058.654180         1      59.0    2.0     2.0      9.0   

       DR1DAY  DR1LANG  DR1MNRSP  DR1HELPD  DBQ095Z  DBD100  DRQSPREP  DR1STY  \
SEQN                                                                            
74781     5.0      1.0       1.0      13.0      1.0     1.0       4.0     2.0   
76977     5.0      1.0       1.0      13.0      1.0     3.0       4.0     1.0   
79530     6.0      1.0       1.0      13.0      1.0     2.0       3.0     2.0   
74171     7.0      1.0       1.0      13.0      2.0     1.0       2.0     2.0   
83046     6.0      1.0       1.0      13.0      4.0     NaN       2.0     2.0   

       DR1SKY  DRQSDIET  DRQSDT1  DRQSDT2  DRQSDT3  DRQSDT4  DRQSDT5  DRQSDT6  \
SEQN                                                                            
74781     NaN       1.0      NaN      NaN      NaN      NaN      NaN      NaN   
76977     1.0       2.0      NaN      NaN      NaN      NaN      NaN      NaN   
79530     NaN       2.0      NaN      NaN      NaN      NaN      NaN      NaN   
74171     NaN       2.0      NaN      NaN      NaN      NaN      NaN      NaN   
83046     NaN       2.0      NaN      NaN      NaN      NaN      NaN      NaN   

       DRQSDT7  DRQSDT8  DRQSDT9  DRQSDT10  DRQSDT11  DRQSDT12  DRQSDT91  \
SEQN                                                                       
74781      7.0      NaN      NaN       NaN       NaN       NaN       NaN   
76977      NaN      NaN      NaN       NaN       NaN       NaN       NaN   
79530      NaN      NaN      NaN       NaN       NaN       NaN       NaN   
74171      NaN      NaN      NaN       NaN       NaN       NaN       NaN   
83046      NaN      NaN      NaN       NaN       NaN       NaN       NaN   

       DR1TNUMF  DR1TKCAL  DR1TPROT  DR1TCARB  DR1TSUGR  DR1TFIBE  DR1TTFAT  \
SEQN                                                                          
74781      20.0    1337.0     35.82    159.57     82.60      13.4     65.13   
76977       7.0    1561.0     75.97    140.12     46.73       7.7     67.21   
79530      14.0    1748.0     47.34    256.44    131.76       9.3     60.71   
74171       7.0     704.0     45.36     95.15     18.92       6.7     14.48   
83046      13.0     902.0     40.82    100.24     49.30       8.5     38.63   

       DR1TSFAT  DR1TMFAT  DR1TPFAT  DR1TCHOL  DR1TATOC  DR1TATOA  DR1TRET  \
SEQN                                                                         
74781    14.519    22.719    23.924     250.0      6.89      0.00    159.0   
76977    21.706    26.472    13.569     251.0      6.98      0.00    180.0   
79530    24.015    23.084    10.053     143.0      4.93      1.27    640.0   
74171     5.736     4.487     2.020      88.0      0.83      0.00     53.0   
83046    14.053    10.871    10.534     115.0      3.45      0.00    379.0   

       DR1TVARA  DR1TACAR  DR1TBCAR  DR1TCRYP  DR1TLYCO  DR1TLZ  DR1TVB1  \
SEQN                                                                       
74781     706.0    2030.0    5500.0      76.0    3012.0  1173.0    0.803   
76977     188.0       5.0      73.0      10.0     157.0   318.0    0.858   
79530     649.0       0.0     110.0       5.0       0.0   255.0    1.317   
74171      59.0       0.0      61.0       5.0       0.0   106.0    0.645   
83046     726.0     131.0    4100.0       2.0    2618.0   336.0    1.265   

       DR1TVB2  DR1TNIAC  DR1TVB6  DR1TFOLA  DR1TFA  DR1TFF  DR1TFDFE  \
SEQN                                                                 

In [10]:
# Pipeline
impute_median = SimpleImputer(strategy='median')
scaler = StandardScaler()
pipe = make_pipeline(impute_median, scaler)

In [11]:
# Fit
pipe.fit(X_train)

Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median')),
                ('standardscaler', StandardScaler())])

In [13]:
# Transform
X_train_tf = pipe.transform(X_train)
X_train_tf.isna().sum().sum()

0

### Model - No PCA

In [14]:
# Fit KNN
knn = KNeighborsClassifier()
knn.fit(X_train_tf, y_train)

KNeighborsClassifier()

In [17]:
%%time
preds = knn.predict(X_test)

ValueError: The feature names should match those that were passed during fit.
Feature names unseen at fit time:
- DRQSDT5


In [16]:
evaluate_classification(knn, X_train_tf,y_train, X_test, y_test)

NameError: name 'X_test_tf' is not defined

### PCA Transform / Model

In [ ]:
# fitting data to preprocessing pipeline
pipe.fit(df)
scaled_df = pipe.transform(df)
scaled_df.isna().sum().sum()

In [ ]:
# Define Target, X, Y
target = 'Diabetes_Risk'
y = scaled_df[target].copy()
X = scaled_df.drop(columns=target).copy()

In [ ]:
# Train/Test Split
X_train_pca,X_test_pca,y_train_pca,y_test_pca = train_test_split(X,y, random_state=321)
X_train_pca.head()

In [ ]:
pca = PCA(n_components=.95)
X_train_pca = pca.fit_transform(X_train_tf)
X_test_pca = pca.transform(X_test_tf)

In [ ]:
pca.n_components_

In [ ]:
# Fit KNN
knn_pca = KNeighborsClassifier()
knn_pca.fit(X_train_pca,y_train)

In [ ]:
%%time
preds_pca = knn_pca.predict(X_test_pca)

In [ ]:
evaluate_classification(knn_pca, X_train_pca,y_train, X_test_pca, y_test)

### Comparing Models

##### How many features were used in the first model, and how many PCs were used in the PCA model?
- 168 in the first model

- 99 in the PCA model

##### Which model performed the best on the test set?
- Both Models performed similarly

##### Which model was the fastest at making predictions?
- In this case they performed in very similar time, with the non-PCA model slightly faster